### ISSUE TITLE
How to specify a custom default value for `JsonIgnoreCondition.WhenWritingDefault`

### ISSUE THREAD

**Issue by ArcticLampyrid** [dotnet/runtime#74326]
How to specify a custom default value for `JsonIgnoreCondition.WhenWritingDefault` ?
```csharp
        // FIXME
        // [DefaultValue(true)]
        // [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
        public bool Foo { get; set; } = true;
```
It seems that `[DefaultValue(true)]` has no effort with `System.Text.Json`, and `WhenWritingDefault` always ignores `false`.

**Comment by krwq** [comment]
Wouldn't that always be `true` for your example?

You can add support for that starting with .NET 7 preview 7 (or targetting .NET 7 preview 7 NuGet package of System.Text.Json):

```csharp
using System.Text.Json;
using System.Text.Json.Serialization.Metadata;

namespace JsonDefaultValueExample
{
    internal class Program
    {
        static void Main(string[] args)
        {
            JsonSerializerOptions options = new()
            {
                TypeInfoResolver = new DefaultJsonTypeInfoResolver()
                {
                    Modifiers = { DefaultValueModifier }
                }
            };

            MyClass obj = new();

            // {"Foo":-1,"EffectivelyAlwaysTrue":true,"StringValue":""}
            Console.WriteLine(JsonSerializer.Serialize(obj, options));
        }

        static void DefaultValueModifier(JsonTypeInfo typeInfo)
        {
            foreach (JsonPropertyInfo property in typeInfo.Properties)
            {
                if (property.AttributeProvider != null)
                {
                    object[] attributes = property.AttributeProvider.GetCustomAttributes(typeof(JsonDefaultValueAttribute), inherit: true);
                    if (attributes.Length == 1) // no multiple attributes so Length is 0 or 1
                    {
                        JsonDefaultValueAttribute attribute = (JsonDefaultValueAttribute)attributes[0];
                        Func<object, object?>? originalGetter = property.Get;

                        if (originalGetter == null)
                        {
                            // you probably want to change exception type here
                            throw new Exception("No getter but default value writing is customized");
                        }

                        // reset any previous JsonIgnore
                        property.ShouldSerialize = (obj, val) => true;

                        Type propertyType = property.PropertyType;

                        object? defaultValue = propertyType.IsValueType ? Activator.CreateInstance(propertyType) : null;
                        property.Get = (object obj) =>
                        {
                            object? value = originalGetter(obj);

                            if ((defaultValue == null && value == null) || (defaultValue != null && defaultValue.Equals(value)))
                            {
                                value = attribute.DefaultValue;
                            }

                            return value;
                        };
                    }
                }
            }
        }
    }

    public class MyClass
    {
        [JsonDefaultValue(-1)]
        public int? Foo { get; set; }

        [JsonDefaultValue(true)]
        public bool EffectivelyAlwaysTrue { get; set; }

        [JsonDefaultValue("")]
        public string? StringValue { get; set; }
    }

    [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property)]
    public class JsonDefaultValueAttribute : Attribute
    {
        public object? DefaultValue { get; set; }

        public JsonDefaultValueAttribute(object? value)
        {
            DefaultValue = value;
        }
    }
}
```

**Comment by krwq** [comment]
FYI: I just updated the sample to be slightly more efficient

**Comment by ArcticLampyrid** [comment]
Wonder why such a general feature is not implemented in 6.0.
It's still too early to switch to .Net 7.0, IMO.
